include/onnx-mlir/Runtime/OMTensor.h

```diff
@@ -0,0 +1,387 @@
+ * presents significant burden for any runtime components trying to interact
+ * with the compiled executable.
+ *
+ * Thus a version of MemRef struct that is amenable to runtime manipulation is
+ * provided as a basis for building any runtime-related components providing
+ * user-facing programming interfaces. All information are dynamically encoded
+ * as members of this struct so that they can be accessed and modified easily
+ * during runtime.
+ *
+ * We will refer to it as an OMTensor.
+ */
+
+struct OMTensor;
+
+#ifndef __cplusplus
+typedef struct OMTensor OMTensor;
+#endif
+
+/* Helper function to compute the number of data elements */
+static inline int64_t getNumOfElems(int64_t *dataSizes, int rank) {
+  int64_t numElem = 1;
+  for (int i = 0; i < rank; i++)
+    numElem *= dataSizes[i];
+  return numElem;
+}
+
+/**
+ * Create a OMTensor with specified data pointer, shape, rank and element type.
+ *
+ * @param data_ptr pointer to tensor data. By default, caller is responsible for
+ * managing the memory this pointer refers to.
+ * @param shape list of integers indicating the tensor shape.
+ * @param rank tensor rank.
+ * @param dtype tensor element data type.
+ * @return pointer to OMTensor created, NULL if creation failed.
+ *
+ */
+OMTensor *omTensorCreate(
+    void *data_ptr, int64_t *shape, int64_t rank, OM_DATA_TYPE dtype);
+
+/**
+ * Create an OMTensor with the specified shape, rank and element type,
+ * allocate uninitialized data for the specified shape.
+ *
+ * @param shape list of integers indicating the tensor shape.
+ * @param rank tensor rank.
+ * @param dtype tensor element data type.
+ * @return pointer to OMTensor created, NULL if creation failed.
+ *
+ */
+OMTensor *omTensorCreateEmpty(int64_t *shape, int64_t rank, OM_DATA_TYPE dtype);
+
+/**
+ * Create an OMTensor with specified data pointer, shape, rank and element type,
+ * manually setting data ptr ownership.
+ *
+ * @param data_ptr pointer to tensor data.
+ * @param shape list of integers indicating the tensor shape.
+ * @param rank tensor rank.
+ * @param dtype tensor element data type.
+ * @param owning whether OMTensor owns the data, if set to true, OMTensor will
+ * release the data_ptr upon destruction.
+ * @return pointer to OMTensor created, NULL if creation failed.
+ *
+ */
+OMTensor *omTensorCreateWithOwnership(void *data_ptr, int64_t *shape,
+    int64_t rank, OM_DATA_TYPE dtype, int owning);
+
+/**
+ * Create an empty OMTensor with specified rank. This constructor returns a
+ * partially filled omTensor; prefer using the new omTensorCreateEmpty()
+ * function to fill shape & stride fields automatically.
+ *
+ * @param rank tensor rank
+ * @return pointer to OMTensor created, NULL if creation failed.
+ *
+ */
+OMTensor *omTensorCreateEmptyDeprecated(int rank);
+
+/**
+ * Destroy the OMTensor struct.
+ *
+ * @param tensor pointer to the OMTensor
+ *
+ */
+void omTensorDestroy(OMTensor *tensor);
+
+/**
```

Comment:
That's a question I do not have answers to, and I presume we will only get answers when we actually get to deal with alignment issues. Current approach is to leave alignment out of the user API design and documentation, because the current assumptions are:
- If data pointer is passed in from an external context, we assume alignment is not an issue (or it has been taken care of, via functions like `_aligned_malloc`). When using `_aligned_malloc`, I don't think it is customary for users to have access to or keep track of both aligned pointers and allocated pointers, so in this sense we don't really have to distinguish between two pointers.
- If data pointer is created by us and we do not have access to things like _aligned_malloc, then we align the data pointer we malloced ourselves and deal with the alignment issue internally and don't involve users in the decision process.

In either case, there is no need for users to know about the potential alignment issue. BTW, the specific function you are referring to is removed now because it is only used by the wrapper code converting MemRef -> OMTensor at the end of inference execution, and I just realized, reading Gong's comment, that I don't need to expose it in the header.